Support Icinga DB

raviks789 opened the pull request:
Add Icinga DB MySQL backend as one of the backends for the monitored nodes.

Resolves #276.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
 {
+    use IcingadbDatabase;
+
+    protected $backend;
```
The member `$backend` is already there. The controller inherits it.
After all, the original code already accesses `$this->backend`.

raviks789 replied:
In the original code we were using the controller from monitoring module, I have changed that to Icinga\Module\Businessprocess\Web\Controller. Here the member $backend is not present. If I use the controller from the monitoring module, it does not work for IcingaDB backend.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
+
+    protected $backend;
+
+    protected $allParams;
```
You don't need that. Accessing params should be done only with `$this->params->getRequired($name)` and `$this->params->get($name[, $default])`.

raviks789 replied:
I have made the changes as discussed in the call.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
+            }
+
+        } else {
+            $this->backend = Backend::createBackend($this->_getParam('backend'));
```
You don't have to change the original code.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
+
     public function showAction()
     {
+        $this->allParams = $this->getAllParams();
```
Remove that line.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
-        $query = $this->backend->select()
-            ->from('hoststatus', array('host_name'))
-            ->where('host_name', $host);
+        if (array_key_exists('backend', $this->allParams)) {
```
Should be `if ($this->params->get('backend') === '_icingadb')`.

raviks789 replied:
I have made the changes as discussed in the call.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
-            ->from('hoststatus', array('host_name'))
-            ->where('host_name', $host);
+        if (array_key_exists('backend', $this->allParams)) {
+            if ($this->allParams['backend'] === '_icingadb') {
```
What happens here if backend is not `_icingadb`?

raviks789 replied:
I have made the changes has discussed in the call.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
-            ->where('host_name', $host);
+        if (array_key_exists('backend', $this->allParams)) {
+            if ($this->allParams['backend'] === '_icingadb') {
+                $this->backend = $this->getDb();
```
Don't use `$this->backend` here. That is the backend from the monitoring controller. Just use `$icingaDb = $this->getDb()` and work with that variable.

lippserd commented on application/controllers/HostController.php:
```diff
@@ -2,21 +2,54 @@
+            $this->params->add('name', $host);
+
+            if ($queryHost !== false) {
+                $this->redirectNow(Url::fromPath('icingadb/host/index')->setParams($this->params));
```
Remove `index` from the URL.

lippserd commented on application/controllers/ProcessController.php:
```diff
@@ -81,11 +82,18 @@ public function showAction()
 
         $this->setTitle($this->translate('Business Process "%s"'), $bp->getTitle());
 
+
```
Remove that change.

lippserd commented on application/controllers/ProcessController.php:
```diff
@@ -169,7 +177,6 @@ protected function prepareRenderer($bp, $node)
             }
             $renderer->setUrl($this->url())
                 ->setPath($this->params->getValues('path'));
-
```
Remove that change.

lippserd commented on application/controllers/ServiceController.php:
```diff
@@ -2,23 +2,56 @@
 {
+    use IcingadbDatabase;
+
+    protected $backend;
```
All required changes to the HostController have to be done here too.

lippserd commented on application/forms/DeleteNodeForm.php:
```diff
@@ -82,7 +82,7 @@ public function setup()
      * @param MonitoringBackend $backend
```
PHPDoc change missing.

lippserd commented on application/forms/EditNodeForm.php:
```diff
@@ -16,9 +18,13 @@ class EditNodeForm extends QuickForm
     /** @var MonitoringBackend */
     protected $backend;
 
+    protected $backendName;
```
PHPDoc missing.

lippserd commented on library/Businessprocess/BpConfig.php:
```diff
@@ -25,9 +28,7 @@ class BpConfig
-     *
-     * @var MonitoringBackend
+     * Backend to retrieve states from
      */
```
Add `@var MonitoringBackend|IcingaDbDatabase` here.

lippserd commented on library/Businessprocess/Common/EnumList.php:
```diff
@@ -0,0 +1,85 @@
+    {
+        return $this->backendName === '_icingadb';
+    }
+}
```
Newline missing. You can configure that in PHPStorm. "Add newline at end of file" or something like that.

lippserd commented on library/Businessprocess/Common/IcingadbDatabase.php:
```diff
@@ -0,0 +1,43 @@
+     */
+    public function getDb()
+    {
+
```
Remove that extra newline.

lippserd commented on library/Businessprocess/IcingaDbBackend.php:
```diff
@@ -0,0 +1,78 @@
+
+    public function fetchHosts()
+    {
+
```
Remove that extra newline.

lippserd commented on library/Businessprocess/IcingaDbBackend.php:
```diff
@@ -0,0 +1,78 @@
+    /** @var BpConfig */
+    protected $config;
+
+    protected $backend;
```
Add PHPDoc and a newline after the member. It makes sense to rename this to `$conn` or `$db` though.

lippserd commented on library/Businessprocess/IcingaDbBackend.php:
```diff
@@ -0,0 +1,78 @@
+    public function fetchHosts()
+    {
+
+        $hosts = Host::on($this->getDb())
```
Use `$this->db` (or whatever variable you've cached in the constructor) instead of the call.

lippserd commented on library/Businessprocess/IcingaDbBackend.php:
```diff
@@ -0,0 +1,78 @@
+
+    public function fetchServices($host)
+    {
+        $query = Service::on($this->backend)
```
Rename `$query` to `$services`. Don't call `fetchAll` and remove the `var_dump`. You also don't need the `assembleSelect()` call. I guess this was only for debugging. Just return `$services` then.

lippserd commented:
> Also, I force-pushed the PR and removed the commits belonging to the other issue.

Could you please fix that again. You have to branch off the master branch.

lippserd reviewed (dismissed):
I think those are the last changes for the moment. Good job 👍

nilmerg reviewed (changes requested):
Please add icingadb as requirement in `module.info` and adjust `doc/02-installation.md` accordingly.

nilmerg commented on library/Businessprocess/HostNode.php:
```diff
@@ -56,8 +58,18 @@ public function getUrl()
 
-        if ($this->getBpConfig()->hasBackendName()) {
-            $params['backend'] = $this->getBpConfig()->getBackendName();
+        if ($this->getBpConfig()->hasBackendName() ||
+            (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend())
+        ) {
+            $backendName = $this->getBpConfig()->getBackendName();
+
+            if (Module::exists('icingadb') &&
+                ($backendName === '_icingadb' || IcingadbSupport::useIcingaDbAsBackend())
+            ) {
+                $params['backend'] = '_icingadb';
+            } else {
+                $params['backend'] = $this->getBpConfig()->getBackendName();
+            }
```
> * Explicit backend?
>   * Add paramter backend

Still doesn't happen. if `$this->getBpConfig()->hasBackendName()` is true **and** `Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()` (on lines 66-67), the real backend name is still ignored.

> * Default backend?
>   * Let the target decide, just like it did the entire time prior this by assuming the default monitoring backend if there is no parameter named backend

Also doesn't happen. Why do you add `_icingadb` as backend? Let the `HostController::showAction` and `ServiceController::showAction` decide what to do if there's no `backend` parameter instead. (i.e. No `backend` parameter? Is Icingadb enabled and the preferred backend? Use it! Monitoring otherwise.)

Note that you don't even need to change anything here for this solution.

nilmerg commented on library/Businessprocess/ServiceNode.php:
```diff
@@ -75,8 +77,18 @@ public function getUrl()
 
-        if ($this->getBpConfig()->hasBackendName()) {
-            $params['backend'] = $this->getBpConfig()->getBackendName();
+        if ($this->getBpConfig()->hasBackendName() ||
+            (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend())
+        ) {
+            $backendName = $this->getBpConfig()->getBackendName();
+
+            if (Module::exists('icingadb') &&
+                ($backendName === '_icingadb' || IcingadbSupport::useIcingaDbAsBackend())
+            ) {
+                $params['backend'] = '_icingadb';
+            } else {
+                $params['backend'] = $this->getBpConfig()->getBackendName();
+            }
```
Same as in my previous comment.

nilmerg commented on application/controllers/HostController.php:
```diff
@@ -2,23 +2,60 @@
 {
+    protected $isIcingadb;
+
+    protected $explicitIcingadb;
```
You shouldn't name this **explicit**Icingadb. It's the exact opposite: *implicit*

Anyway, what about the actual meaning: `$isIcingadbPreferred`

And add documentation!

nilmerg commented on application/controllers/ServiceController.php:
```diff
@@ -2,26 +2,71 @@
 {
+    protected $isIcingadb;
+
+    protected $explicitIcingadb;
```
Same as in my previous comment.

nilmerg commented on application/controllers/ServiceController.php:
```diff
@@ -2,26 +2,71 @@
 
 namespace Icinga\Module\Businessprocess\Controllers;
 
+use Dompdf\Exception;
```
Remove this.

nilmerg commented on library/Businessprocess/BpConfig.php:
```diff
@@ -290,9 +292,23 @@ public function setBackend(MonitoringBackend $backend)
+
+                $this->backend = IcingaDbObject::fetchDb();
+            } else {
+                if (! Module::exists('monitoring') && Module::exists('icingadb')) {
+                    $this->backend = IcingaDbObject::fetchDb();
```
This is redundant. `IcingadbSupport::useIcingaDbAsBackend()` is already true if monitoring is disabled.

nilmerg commented on library/Businessprocess/State/IcingaDbState.php:
```diff
@@ -0,0 +1,151 @@
+    /** @var BpConfig */
+    protected $config;
+
+    /** @var IcingadbDatabase */
```
That's a trait, and not true after all.

nilmerg reviewed (changes requested):
* Please remove all traces of `getBackendName() === '_icingadb'` and its other variations.
* Extend all cases of `Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()` with a check for `$config->hasBackendName()` (if applicable, i.e. Is Icingadb preferred and config has no backend name? Use Icingadb. Monitoring otherwise.)

* Provide native implementations for the following hooks: `icingadb/HostActions` and `icingadb/ServiceActions`

nilmerg commented:
Ah, **and**: Please add icingadb as requirement in `module.info` and adjust `doc/02-installation.md` accordingly.

nilmerg commented on application/controllers/ServiceController.php:
```diff
@@ -2,26 +2,67 @@
+
+    protected function moduleInit()
+    {
+        $this->isIcingadbPreferred = Module::exists('icingadb')
+            && IcingadbSupport::useIcingaDbAsBackend();
```
Add `! $this->params->has('backend')`

nilmerg commented on application/controllers/HostController.php:
```diff
@@ -2,23 +2,59 @@
 
 class HostController extends Controller
 {
+    /** @var bool*/
+    /** True if business process prefers to use icingadb as backend for it's nodes */
```
```suggestion
    /**
     * True if business process prefers to use icingadb as backend for it's nodes
     *
     * @var bool
     */
```

nilmerg commented on application/controllers/HostController.php:
```diff
@@ -2,23 +2,59 @@
+
+    protected function moduleInit()
+    {
+        $this->isIcingadbPreferred = Module::exists('icingadb')
+            && IcingadbSupport::useIcingaDbAsBackend();
```
Add `! $this->params->has('backend')`

nilmerg commented on application/clicommands/ProcessCommand.php:
```diff
@@ -132,7 +135,14 @@ public function checkAction()
             $node = $bp->getNode($nodeName);
-            MonitoringState::apply($bp);
+            if (Module::exists('icingadb') &&
+                ($bp->getBackendName() === '_icingadb' || IcingadbSupport::useIcingaDbAsBackend())
```
Remove `$bp->getBackendName() === '_icingadb'`.
Add `! $bp->hasBackendName()`.

nilmerg reviewed (changes requested):
## Having an explicit backend configured shows this

### when clicking on the "Bussiness Impact" link in an object's detail

```
Uncaught TypeError: ipl\Orm\Model::on(): Argument #1 ($db) must be of type ipl\Sql\Connection, Icinga\Module\Monitoring\Backend\Ido\IdoBackend given, called in /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php on line 65 and defined in /usr/local/src/ipl-orm/src/Model.php:70
Stack trace:
#0 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(65): ipl\Orm\Model::on(Object(Icinga\Module\Monitoring\Backend\Ido\IdoBackend))
#1 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(42): Icinga\Module\Businessprocess\State\IcingaDbState->reallyRetrieveStatesFromBackend()
#2 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(32): Icinga\Module\Businessprocess\State\IcingaDbState->retrieveStatesFromBackend()
#3 /usr/share/icingaweb2-modules/businessprocess/application/controllers/NodeController.php(91): Icinga\Module\Businessprocess\State\IcingaDbState::apply(Object(Icinga\Module\Businessprocess\BpConfig))
#4 /icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Businessprocess\Controllers\NodeController->impactAction()
#5 /icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch('impactAction')
#6 /icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#7 /icingaweb2/library/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#8 /icingaweb2/library/Icinga/Application/webrouter.php(109): Icinga\Application\Web->dispatch()
#9 /icingaweb2/public/index.php(4): require_once('/icingaweb2/lib...')
#10 {main}
thrown
```

### when opening the process

```
Uncaught TypeError: ipl\Orm\Model::on(): Argument #1 ($db) must be of type ipl\Sql\Connection, Icinga\Module\Monitoring\Backend\Ido\IdoBackend given, called in /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php on line 65 and defined in /usr/local/src/ipl-orm/src/Model.php:70
Stack trace:
#0 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(65): ipl\Orm\Model::on(Object(Icinga\Module\Monitoring\Backend\Ido\IdoBackend))
#1 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(42): Icinga\Module\Businessprocess\State\IcingaDbState->reallyRetrieveStatesFromBackend()
#2 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(32): Icinga\Module\Businessprocess\State\IcingaDbState->retrieveStatesFromBackend()
#3 /usr/share/icingaweb2-modules/businessprocess/application/controllers/ProcessController.php(90): Icinga\Module\Businessprocess\State\IcingaDbState::apply(Object(Icinga\Module\Businessprocess\BpConfig))
#4 /icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Businessprocess\Controllers\ProcessController->showAction()
#5 /icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch('showAction')
#6 /icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#7 /icingaweb2/library/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#8 /icingaweb2/library/Icinga/Application/webrouter.php(109): Icinga\Application\Web->dispatch()
#9 /icingaweb2/public/index.php(4): require_once('/icingaweb2/lib...')
#10 {main}
thrown
```

raviks789 commented:
> ## Having an explicit backend configured shows this
> ### when clicking on the "Bussiness Impact" link in an object's detail
> ```
> Uncaught TypeError: ipl\Orm\Model::on(): Argument #1 ($db) must be of type ipl\Sql\Connection, Icinga\Module\Monitoring\Backend\Ido\IdoBackend given, called in /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php on line 65 and defined in /usr/local/src/ipl-orm/src/Model.php:70
> Stack trace:
> #0 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(65): ipl\Orm\Model::on(Object(Icinga\Module\Monitoring\Backend\Ido\IdoBackend))
> #1 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(42): Icinga\Module\Businessprocess\State\IcingaDbState->reallyRetrieveStatesFromBackend()
> #2 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(32): Icinga\Module\Businessprocess\State\IcingaDbState->retrieveStatesFromBackend()
> #3 /usr/share/icingaweb2-modules/businessprocess/application/controllers/NodeController.php(91): Icinga\Module\Businessprocess\State\IcingaDbState::apply(Object(Icinga\Module\Businessprocess\BpConfig))
> #4 /icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Businessprocess\Controllers\NodeController->impactAction()
> #5 /icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch('impactAction')
> #6 /icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
> #7 /icingaweb2/library/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
> #8 /icingaweb2/library/Icinga/Application/webrouter.php(109): Icinga\Application\Web->dispatch()
> #9 /icingaweb2/public/index.php(4): require_once('/icingaweb2/lib...')
> #10 {main}
> thrown
> ```
> 
> ### when opening the process
> ```
> Uncaught TypeError: ipl\Orm\Model::on(): Argument #1 ($db) must be of type ipl\Sql\Connection, Icinga\Module\Monitoring\Backend\Ido\IdoBackend given, called in /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php on line 65 and defined in /usr/local/src/ipl-orm/src/Model.php:70
> Stack trace:
> #0 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(65): ipl\Orm\Model::on(Object(Icinga\Module\Monitoring\Backend\Ido\IdoBackend))
> #1 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(42): Icinga\Module\Businessprocess\State\IcingaDbState->reallyRetrieveStatesFromBackend()
> #2 /usr/share/icingaweb2-modules/businessprocess/library/Businessprocess/State/IcingaDbState.php(32): Icinga\Module\Businessprocess\State\IcingaDbState->retrieveStatesFromBackend()
> #3 /usr/share/icingaweb2-modules/businessprocess/application/controllers/ProcessController.php(90): Icinga\Module\Businessprocess\State\IcingaDbState::apply(Object(Icinga\Module\Businessprocess\BpConfig))
> #4 /icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Businessprocess\Controllers\ProcessController->showAction()
> #5 /icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch('showAction')
> #6 /icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
> #7 /icingaweb2/library/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
> #8 /icingaweb2/library/Icinga/Application/webrouter.php(109): Icinga\Application\Web->dispatch()
> #9 /icingaweb2/public/index.php(4): require_once('/icingaweb2/lib...')
> #10 {main}
> thrown
> ```

This should be fixed now.

raviks789 commented:
> Yeah, but I don't know what you've changed now. I personally tracked this down to two remaining cases of checks for `_icingadb` as backend name. Please re-check if that might be the only required change to fix these errors.

Thanks for bringing that to my notice as well. I have removed  two remaining cases of checks for `_icingadb` as backend name. The other change was, I have changed `$this->backend = $config->getBackend()` to `$this->backend = IcingaDbObject::fetchDb()` in IcingaDbState ([library/Businessprocess/State/IcingaDbState.php](https://github.com/Icinga/icingaweb2-module-businessprocess/commit/df3f363f1b4164f412d6f81853fb4aca228434c6#diff-ce855daf8aeb62055e71300301653c7f21e01d6206d76dab5d2faea1c4312177)).
